### PR TITLE
free(socket_path) to suppress leak sanitizer false-positive with -O2

### PR DESCRIPTION
**Pull request by stapelberg**
related to #4086 (triggered by our meson config)
related to #4087 (sanitizer cleanup tracking bug)

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -376,6 +376,11 @@ int main(int argc, char *argv[]) {
                     char *socket_path = root_atom_contents("I3_SOCKET_PATH", NULL, 0);
                     if (socket_path) {
                         printf("%s\n", socket_path);
+                        // With -O2 (i.e. the buildtype=debugoptimized meson
+                        // option, which we set by default), gcc 9.2.1 optimizes
+                        // away socket_path at this point, resulting in a Leak
+                        // Sanitizer report. An explicit free helps:
+                        free(socket_path);
```
Comment (orestisfl) [inline src/main.c:383]: I don't understand your explanation here. It seems that this is a leak because we have to free the results of `root_atom_contents`. So, instead I propose to free them in both paths.

**Review by orestisfl — CHANGES_REQUESTED** [review]

File: src/main.c (hunk 1 above)
Comment (stapelberg) [reply src/main.c:383]: What do you mean by both paths? This path is `socket_path != NULL`, so it doesn’t make sense to free in the other part, where `socket_path` is already `NULL`?

Comment (orestisfl) [reply src/main.c:383]: Sorry, it was late.

Anyway, this is not a false-positive, right? We are supposed to free `socket_path`.

Comment (stapelberg) [reply src/main.c:383]: No worries :)

Well, we terminate the process right after, so the OS will free that memory. Leaving cleanup to the OS is a widely used strategy.

The nuance here revolves around whether `socket_path` is reachable at process exit or not. When building with `buildtype=debug` (-g), `socket_path` is still considered reachable at `exit` time, which matches the intuitive interpretation: the variable is still in scope, so the memory should be reachable.

When building with `buildtype=debugoptimized` (-g -O2), `socket_path` is optimized out so aggressively that the leak sanitizer cannot find it and hence must consider it unreachable at `exit` time.

It seems possible to me that future versions of the leak sanitizer can deal with this better, or compiler optimizations might be relaxed when the leak sanitizer is used. For now, we just do the extra `free(socket_path)`, which is neither incorrect, nor required (under normal operation).

Comment (orestisfl) [reply src/main.c:383]: Ok I see. I think that freeing before exit is good either way so I don't think we should justify it with so much text (harder on the reader) but if you want to keep the comment, use `/*` for consistency.

Comment (stapelberg) [reply src/main.c:383]: I’d like to keep the comment, because an explicit free is at least an inconsistency. Without the comment, contributors might assume that all memory always needs to be freed before exiting the process (which is laborious, error-prone and unnecessary), or go the other way and remove the explicit `free()` for consistency, without realizing that it will break further down the road.

Updated the comment marks.

**Review by orestisfl — APPROVED** [review]
